accounts/accounts/controllers/registration.py

```diff
@@ -0,0 +1,303 @@
+        except Exception as e:
+            data['error'] = 'Could not save user profile; please try again'
+            return data, status.HTTP_500_INTERNAL_SERVER_ERROR, {}
+
+        # We need a new session, to update user's data.
+        _logout(session.session_id)
+        new_session, new_cookie = _login(user, auth, ip)
+        data.update({'cookies': {
+            'session_cookie': (new_cookie, new_session.expires)
+        }})
+        return data, status.HTTP_303_SEE_OTHER, {}
+    return data, status.HTTP_200_OK, {}
+
+
+class ProfileForm(Form):
+    """User registration form."""
+
+    COUNTRIES = [('', '')] + \
+        [(country.alpha_2, country.name) for country in pycountry.countries]
+    RANKS = [('', '')] + domain.UserProfile.RANKS
+    GROUPS = [
+        (key, group['name']) for key, group in taxonomy.GROUPS.items()
+        if not group.get('is_test', False)
+    ]
+    CATEGORIES = [
+        (archive['name'], [
+            (category_id, category['name'])
+            for category_id, category in taxonomy.CATEGORIES_ACTIVE.items()
+            if category['in_archive'] == archive_id
+        ])
+        for archive_id, archive in taxonomy.ARCHIVES_ACTIVE.items()
+    ]
+    """Categories grouped by archive."""
+
+    user_id = HiddenField('User ID')
+
+    email = StringField(
+        'Email address',
+        validators=[Email(), Length(max=255), DataRequired()],
+        description="You must be able to receive mail at this address."
+        " We take <a href='https://arxiv.org/help/email-protection'>"
+        " strong measures</a> to protect your email address from viruses and"
+        " spam. Do not enter an e-mail address that belongs to someone"
+        " else: if we discover that you've done so, we will suspend your"
+        " account."
+    )
+
+    username = StringField('Username',
+                           validators=[Length(min=5, max=20), DataRequired()])
+    forename = StringField('First or given name',
+                           validators=[Length(min=1, max=50), DataRequired()])
+    surname = StringField('Last or family name',
+                          validators=[Length(min=1, max=50), DataRequired()])
+    suffix = StringField('Suffix', validators=[Length(max=50)])
+    organization = StringField(
+        'Organization',
+        validators=[Length(max=255), DataRequired()],
+        description='This field accepts '
+                    '<a href="https://arxiv.org/user/tex_accents">'
+                    'pidgin TeX (\\\'o)</a> for foreign characters.'
+    )
+    country = SelectField('Country', choices=COUNTRIES,
+                          validators=[DataRequired()])
+    status = SelectField('Status', choices=RANKS, validators=[DataRequired()])
+
+    groups = MultiCheckboxField('Group(s) to which you would like to submit',
+                                choices=GROUPS, default='')
+    default_category = OptGroupSelectField('Your default category',
```

Comment:
They way the category dropdown is being generated is quite faithful to the classic form, so this is not completely germane to the review. Just want to put in my $.02 for future improvements:
* we should probably just subsume the Physics "archive categories" (e.g. hep-th, hep-ph) into Physics rather than including them as solo members under headings of the same name.
* we should consider adding the category ID now that we have some duplicate category names (e.g. "Machine Learning (cs.LG)")
* typing the category name + autocomplete was kind of nice (from classic form)